server_src/runners/history.js

```diff
@@ -25,15 +23,284 @@ const script = require("./scriptCheck.js");
-const { downloadFromOctoPrint } = require("../utils/download.util");
 
 const { writePoints } = require("../lib/influxExport.js");
 
-const logger = new Logger("OctoFarm-HistoryCollection");
 let counter = 0;
 let errorCounter = 0;
 
+const {
+  generateHistoryIndex,
+  generateHistoryState,
+  generateHistoryReason,
+  generateHistoryFileInformation,
+  generateHistoryDate,
+  generateHistoryPrintTime,
+  generateHistoryResendStatistics,
+  generateHistoryJobStatistics,
+} = require("../services/history.service.js");
+
+const { definePrinterName } = require("../services/printer.service.js");
+
+const {
+  checkForValidPrinterCostSettings,
+} = require("../services/printer-settings.service.js");
+
+const {
+  octoprintGrabThumbnail,
+} = require("../services/octoprint/octoprint-plugins.service.js");
+
+const {
+  objectIDValidator,
+  generateObjectId,
+} = require("../utils/database.utils.js");
+
+class History {
+  _id;
+  historyIndex;
+  printerName;
+  printerID;
+  costSettings;
+  success;
+  reason;
+  fileName;
+  fileDisplay;
+  filePath;
+  startDate;
+  endDate;
+  thumbnail;
+  printTime;
+  filamentSelection;
+  previousFilamentSelection;
+  job;
+  notes;
+  snapshot;
+  resends;
+
+  // Function to create a new history record
+  // Will be able to cut this down when state.js refactor
+  async createNewHistoryRecord({
+    printer = undefined,
+    state = undefined,
+    payload = undefined,
+    job = undefined,
+    files = undefined,
+    resendStats = undefined,
+  }) {
+    this._id = await generateObjectId();
+    this.historyIndex = await generateHistoryIndex();
+    this.printerName = await definePrinterName(
+      printer.printerURL,
+      printer.settingsApperance.name
+    );
+
+    this.printerID = await objectIDValidator(printer._id);
+    this.costSettings = await checkForValidPrinterCostSettings(
+      printer.costSettings
+    );
+    this.success = await generateHistoryState(state);
+    this.reason = await generateHistoryReason(payload.reason);
+    this.fileName = await generateHistoryFileInformation("name", payload);
+    this.fileDisplay = await generateHistoryFileInformation("display", payload);
+    this.filePath = await generateHistoryFileInformation("path", payload);
+    this.startDate = await generateHistoryDate("start-date", payload.time);
+    this.endDate = await generateHistoryDate("end-date", payload.time);
+    this.thumbnail = await octoprintGrabThumbnail(
+      state,
+      files,
+      payload.name,
+      this._id,
+      printer.printerURL,
+      printer.apikey
+    );
+    this.printTime = await generateHistoryPrintTime(payload.time);
+    this.filamentSelection;
+    this.previousFilamentSelection;
+    this.job = await generateHistoryJobStatistics(job);
+    // There are no notes on creation
+    this.notes = "";
+    this.snapshot;
+    this.resends = await generateHistoryResendStatistics(resendStats);
+    this.timelapse;
+    console.log(this);
```

Comment:
Debug is so much nicer
![image](https://user-images.githubusercontent.com/6005355/116816842-f2e2e480-ab63-11eb-8b12-20e46f47d895.png)